### PR TITLE
RAC-118: refactor: make QuantifiedAssociations immutable

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

This PR contain refactor of function merge in QuantifiedAssociations class. 
This function is now immutable. 

This use array_replace_recursive instead of using several loops 

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Done
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/QuantifiedAssociation/QuantifiedAssociations.php [hunk 1]
```diff
@@ -249,39 +228,21 @@ function (QuantifiedLink $quantifiedLink) use ($grantedProductModelCodes) {
...
+    private function normalizeWithIndexedIdentifiers()
+    {
+        $quantifiedAssociationsNormalized = $this->normalize();
 
-        if (empty($matchingSourceQuantifiedAssociations)) {
-            return null;
+        $result = [];
+        foreach ($quantifiedAssociationsNormalized as $associationType => $associationsNormalized) {
+            foreach ($associationsNormalized as $quantifiedLinksType => $quantifiedLinksNormalized) {
+                $result[$associationType][$quantifiedLinksType] = array_column(
```
Comment (tseho) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/QuantifiedAssociation/QuantifiedAssociations.php:238]: I know the trick with array_column is neat but personally, it's always harder to read.

**Comment by tseho** [comment]
This is so much better already. The merger should never have been mutable in the first place.

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Adder/QuantifiedAssociationFieldAdder.php [hunk 2]
```diff
@@ -41,6 +41,11 @@ public function __construct(
      */
     public function addFieldData($product, $field, $data, array $options = [])
     {
-        $product->getQuantifiedAssociations()->merge(QuantifiedAssociations::createFromNormalized($data));
+        $currentQuantifiedAssociations = $product->getQuantifiedAssociations();
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Adder/QuantifiedAssociationFieldAdder.php:44]: maybe this should be done in the trait? I know that it's should be done in the next PR but maybe we can do it in this one

Comment (StevenVAIDIE) [reply src/Akeneo/Pim/Enrichment/Component/Product/Updater/Adder/QuantifiedAssociationFieldAdder.php]: Done here : https://github.com/akeneo/pim-community-dev/pull/12427/commits/a6adacec0d1d7b7b6195afda957cb398b9333420

**Pull request merged**